core/src/test/scala/unit/kafka/server/OffsetFetchRequestTest.scala

```diff
@@ -0,0 +1,227 @@
+
+  override def brokerCount: Int = 1
+
+  val brokerId: Integer = 0
+  val offset = 15L
+  val leaderEpoch: Optional[Integer] = Optional.of(3)
+  val metadata = "metadata"
+
+  override def brokerPropertyOverrides(properties: Properties): Unit = {
+    properties.put(KafkaConfig.BrokerIdProp, brokerId.toString)
+    properties.put(KafkaConfig.OffsetsTopicPartitionsProp, "1")
+    properties.put(KafkaConfig.OffsetsTopicReplicationFactorProp, "1")
+    properties.put(KafkaConfig.TransactionsTopicPartitionsProp, "1")
+    properties.put(KafkaConfig.TransactionsTopicReplicationFactorProp, "1")
+    properties.put(KafkaConfig.TransactionsTopicMinISRProp, "1")
+  }
+
+  @BeforeEach
+  override def setUp(): Unit = {
+    doSetup(createOffsetsTopic = false)
+
+    TestUtils.createOffsetsTopic(zkClient, servers)
+  }
+
+  @Test
+  def testOffsetFetchRequestLessThanV8(): Unit = {
+    val topic = "topic"
+    createTopic(topic)
+
+    val groupId = "groupId"
+    val tpList = singletonList(new TopicPartition(topic, 0))
+    val topicOffsets = tpList.asScala.map{
+      tp => (tp, new OffsetAndMetadata(offset, leaderEpoch, metadata))
+    }.toMap.asJava
+
+    consumerConfig.setProperty(ConsumerConfig.GROUP_ID_CONFIG, groupId)
+    val consumer = createConsumer()
+    consumer.assign(tpList)
+    consumer.commitSync(topicOffsets)
+    consumer.close()
+    // testing from version 1 onward since version 0 read offsets from ZK
+    for (version <- 1 to ApiKeys.OFFSET_FETCH.latestVersion()) {
+      if (version < 8) {
+        val request =
+          if (version < 7) {
+            new OffsetFetchRequest.Builder(
+              groupId, false, tpList, false)
+              .build(version.asInstanceOf[Short])
+          } else {
+            new OffsetFetchRequest.Builder(
+              groupId, false, tpList, true)
+              .build(version.asInstanceOf[Short])
+          }
+        val response = connectAndReceive[OffsetFetchResponse](request)
+        val topicData = response.data().topics().get(0)
+        val partitionData = topicData.partitions().get(0)
+        if (version < 3) {
+          assertEquals(AbstractResponse.DEFAULT_THROTTLE_TIME, response.throttleTimeMs())
+        }
+        assertEquals(Errors.NONE, response.error())
+        assertEquals(topic, topicData.name())
+        assertEquals(0, partitionData.partitionIndex())
+        assertEquals(offset, partitionData.committedOffset())
+        if (version >= 5) {
+          // committed leader epoch introduced with V5
+          assertEquals(leaderEpoch.get(), partitionData.committedLeaderEpoch())
+        }
+        assertEquals(metadata, partitionData.metadata())
+        assertEquals(Errors.NONE.code(), partitionData.errorCode())
+      }
+    }
+  }
+
+  @Test
+  def testOffsetFetchRequestV8AndAbove(): Unit = {
+    val groupOne = "group1"
+    val groupTwo = "group2"
+    val groupThree = "group3"
+    val groupFour = "group4"
+    val groupFive = "group5"
```

Comment:
If we move the single group case to the previous test, we can make this one work on a collection of groups. Instead of named groups like `groupOne`, we can just use group(1).

Reply:
I refactored this test class a bit to make use of more helper methods - is there something additional we can do here to make it work on a collection of groups? Are you suggesting adding an additional method that can create some number of groups that we can use the test with?

Reply:
We can do:
```
val groups = (0 until 5).map(i => s"group$i")
```
And use groups(0), groups(1) etc. instead of groupOne, groupTwo.